cmd/addon/update.go

```diff
@@ -19,51 +22,64 @@ var (
-	updateCmd.Flags().StringVar(&addonFlags.Manifest, "manifest", addonFlags.Manifest, "The manifest file content of add-on, need to be base64 encode")
+	updateCmd.Flags().StringVarP(&addonFlags.FromFile, "from", "f", addonFlags.FromFile, "The manifest file path of add-on")
 	updateCmd.Flags().StringToStringVar(&addonFlags.Values, "set", addonFlags.Values, "Set value to replace parameters defined in manifest")
+	updateCmd.Flags().BoolVar(&addonFlags.Override, "override", addonFlags.Override, "Overriding the current add-on configuration, default to false as merge mode")
 }
 
 func UpdateCmd() *cobra.Command {
+	updateCmd.PreRunE = func(cmd *cobra.Command, args []string) error {
+		if addonFlags.Override &&
+			addonFlags.FromFile == "" {
+			return errors.New("must set addon manifest by -f <file-path> when update type is override")
+		}
+		if addonFlags.FromFile != "" && !utils.IsFileExists(addonFlags.FromFile) {
+			return fmt.Errorf("manifest file %s is not exist", addonFlags.FromFile)
+		}
+		return nil
+	}
 	updateCmd.Run = func(cmd *cobra.Command, args []string) {
 		name := args[0]
 		addon, err := common.DefaultDB.GetAddon(name)
 		if err != nil {
 			logrus.Fatalln(err)
 		}
-		var manifestFile string
-		manifestContent := string(addon.Manifest)
+		var manifestString string
+		isChanged := false
 		if addonFlags.FromFile != "" {
-			manifestFile = addonFlags.FromFile
-		} else if addonFlags.Manifest != "" {
-			manifestContent = addonFlags.Manifest
-		}
-
-		manifestString, err := common.GetManifest(manifestFile, manifestContent)
-		if err != nil {
-			logrus.Fatalln(err)
+			manifestString, err = common.GetManifest(addonFlags.FromFile, "")
+			if err != nil {
+				logrus.Error(err)
+				return
+			}
 		}
-		manifest := []byte(manifestString)
-
-		isChanged := false
-		if addonFlags.Description != "" && addonFlags.Description != addon.Description {
+		if addonFlags.Override {
+			addon.Manifest = []byte(manifestString)
 			addon.Description = addonFlags.Description
-			isChanged = true
-		}
-
-		if !bytes.Equal(manifest, addon.Manifest) {
-			addon.Manifest = manifest
-			isChanged = true
-		}
-
-		if !reflect.DeepEqual(addonFlags.Values, addon.Values) {
 			addon.Values = addonFlags.Values
 			isChanged = true
+		} else {
+			if manifestString != "" {
+				addon.Manifest = []byte(manifestString)
+				isChanged = true
+			}
+			if addonFlags.Description != "" && addonFlags.Description != addon.Description {
+				addon.Description = addonFlags.Description
+				isChanged = true
+			}
+			if addon.Values == nil {
+				addon.Values = make(types.StringMap)
+			}
+			for k, v := range addonFlags.Values {
+				addon.Values[k] = v
+			}
+			if !reflect.DeepEqual(addonFlags.Values, addon.Values) {
+				isChanged = true
+			}
 		}
-
 		if isChanged {
 			if err := common.DefaultDB.SaveAddon(addon); err != nil {
-				logrus.Fatalln(err)
+				logrus.Error(err)
```

Comment:
Since it's using `updateCmd.Run` here, will it quit the command if here's an error occurs using Error instead of Fatal?

Reply:
It will quit command as this is the end of the function. I am try to avoid using `Fatal` in command function. For now, using `Error` or `Fatal` has no difference here.